File: validators/PCA/pcas_tsv/validate.R
#!/usr/bin/env Rscript

args <- commandArgs(trailingOnly = TRUE)

validate_file <- function(path) {
  ok <- TRUE

  tryCatch({
    df <- read.table(path, header = TRUE, sep = "\t", check.names = FALSE)

    if (colnames(df)[1] != "PC1") {
      stop(sprintf("first column must be 'PC1', got '%s'", colnames(df)[1]))
    }
    if (nrow(df) == 0) stop("empty file")
    if (ncol(df) < 10) stop("too few columns")
    if (anyNA(df[1, ])) stop("first data row contains non-numeric values in PC columns")

  }, error = function(e) {
    message(sprintf("FAIL\t%s\t%s", path, e$message))
    ok <<- FALSE
  })

  if (ok) {
    message(sprintf("OK\t%s", path))
  }
  ok
}

# pass iff every file validates; exit non-zero on any failure (omnibenchmark
# `validate outputs` keys pass/fail off the exit code).
results <- vapply(args, validate_file, logical(1))
if (!all(results)) quit(status = 1)

File: validators/DATA/rawdata_h5ad/validate.py
#!/usr/bin/env python
"""Validate a DATA stage .h5ad.

The downstream R modules read obs columns (batch/sample/labels) with
rhdf5::h5read, which expects a flat array. anndata writes string columns as
categorical *groups* (categories + codes), which h5read mangles into a
degenerate vector. Fail if any obs column is categorical-encoded.
"""
import sys

import h5py


def validate_file(path):
    try:
        with h5py.File(path, "r") as f:
            bad = [c for c in f["obs"]
                   if f["obs"][c].attrs.get("encoding-type") == "categorical"]
            if bad:
                raise ValueError(
                    "obs columns are categorical-encoded (must be flat "
                    f"string-arrays for rhdf5::h5read): {', '.join(bad)}")
    except Exception as e:
        print(f"FAIL\t{path}\t{e}")
        return False
    print(f"OK\t{path}")
    return True


# pass iff every file validates; exit non-zero on any failure (omnibenchmark
# `validate outputs` keys pass/fail off the exit code).
ok = all([validate_file(arg) for arg in sys.argv[1:]])
sys.exit(0 if ok else 1)
